src/Psalm/Internal/Type/TypeExpander.php

```diff
@@ -878,4 +849,85 @@ private static function expandConditional(
 
         return [$return_type];
     }
+
+    /**
+     * @param TKeyOfArray|TValueOfArray $return_type
+     * @return non-empty-list<Atomic>
+     */
+    private static function expandKeyOfValueOfArray(
+        Codebase $codebase,
+        $return_type,
+        ?string $self_class,
+        bool $evaluate_class_constants
+    ): array {
+        $type_param = $return_type->type;
+
+        if ($evaluate_class_constants && $type_param instanceof TClassConstant) {
+            if ($type_param->fq_classlike_name === 'self' && $self_class) {
+                $type_param->fq_classlike_name = $self_class;
+            }
+
+            if (!$codebase->classOrInterfaceExists($type_param->fq_classlike_name)) {
+                return [$return_type];
+            }
+
+            try {
+                $type_param = $codebase->classlikes->getClassConstantType(
+                    $type_param->fq_classlike_name,
+                    $type_param->const_name,
+                    ReflectionProperty::IS_PRIVATE
+                );
+            } catch (CircularReferenceException $e) {
+                return [$return_type];
+            }
+            if (!$type_param) {
+                return [$return_type];
+            }
+        }
+
+        if (!$type_param instanceof Union) {
+            $type_param = new Union([$type_param]);
+        }
+
+        // Merge keys/values of provided array types
+        $new_return_types = [];
+        foreach ($type_param->getAtomicTypes() as $type_atomic) {
+            // Abort if any type of the param's union is invalid
+            if (!$type_atomic instanceof TKeyedArray
+                && !$type_atomic instanceof TArray
+                && !$type_atomic instanceof TList
+            ) {
+                break;
+            }
+
+            // Transform all types to TArray if needed
+            if ($type_atomic instanceof TList) {
+                $type_atomic = new TArray([
+                    new Union([new TInt()]),
+                    $type_atomic->type_param
+                ]);
+            }
+            if ($type_atomic instanceof TKeyedArray) {
+                $type_atomic = $type_atomic->getGenericArrayType();
+            }
+
+            // Add key-of/value-of type to return types list
+            if ($return_type instanceof TKeyOfArray) {
+                $new_return_types = array_merge(
+                    $new_return_types,
+                    array_values($type_atomic->type_params[0]->getAtomicTypes())
+                ) ;
+            } else {
+                $new_return_types = array_merge(
+                    $new_return_types,
+                    array_values($type_atomic->type_params[1]->getAtomicTypes())
+                ) ;
+            }
+        }
+
+        if (empty($new_return_types)) {
```

Comment:
please avoid empty, `if(new_return_types !== [])` is more clear

Reply:
Done